File: backend/services/chat_service.py
"""Chat orchestration — connects frontend messages to Anthropic API + skill scripts."""
import json
import tempfile
import time
from collections.abc import AsyncIterator
from pathlib import Path
from typing import Any

from ..config import SKILL_MD_CONTENT
from . import anthropic_client, script_runner, session_manager


def build_tool_definitions() -> list[dict[str, Any]]:
    """Build Anthropic tool definitions matching interview_session subcommands."""
    return [
        {
            "name": "read_session_state",
            "description": "Read current interview session state including runtime_status, active_stage, current_question, config, progress, and available commands. Call this first to understand where the interview is.",
            "input_schema": {"type": "object", "properties": {}, "required": []},
        },
        {
            "name": "interview_session_start",
            "description": "Start the interview — move from CONFIG_READY to RUNNING with the first question.",
            "input_schema": {"type": "object", "properties": {}, "required": []},
        },
        {
            "name": "interview_session_next",
            "description": "Advance to the next question in the current stage.",
            "input_schema": {"type": "object", "properties": {}, "required": []},
        },
        {
            "name": "interview_session_hint",
            "description": "Increase the hint level for the current question. Hints range from 1 (direction) to 5 (full solution).",
            "input_schema": {"type": "object", "properties": {}, "required": []},
        },
        {
            "name": "interview_session_repeat",
            "description": "Repeat the current question to the candidate.",
            "input_schema": {"type": "object", "properties": {}, "required": []},
        },
        {
            "name": "interview_session_explain",
            "description": "Explain the current question's intent and what a good answer should cover.",
            "input_schema": {"type": "object", "properties": {}, "required": []},
        },
        {
            "name": "interview_session_skip",
            "description": "Skip the current question and record it as skipped.",
            "input_schema": {
                "type": "object",
                "properties": {
                    "feedback": {"type": "string", "description": "Brief reason for skipping"}
                },
                "required": [],
            },
        },
        {
            "name": "interview_session_pause",
            "description": "Pause the interview at the current question. State changes to PAUSED.",
            "input_schema": {"type": "object", "properties": {}, "required": []},
        },
        {
            "name": "interview_session_continue",
            "description": "Continue a paused interview from the current question.",
            "input_schema": {"type": "object", "properties": {}, "required": []},
        },
        {
            "name": "interview_session_score",
            "description": "Generate a mid-session score snapshot showing current performance without ending the interview.",
            "input_schema": {"type": "object", "properties": {}, "required": []},
        },
        {
            "name": "interview_session_report",
            "description": "Generate the final structured interview evaluation report. Only call when the interview stage sequence is complete or the candidate explicitly requests it.",
            "input_schema": {"type": "object", "properties": {}, "required": []},
        },
        {
            "name": "interview_session_configure",
            "description": "Update interview configuration: role, mode, strength, tone, level, or focus.",
            "input_schema": {
                "type": "object",
                "properties": {
                    "mode": {"type": "string", "description": "Interview mode: 完整模拟, 项目深挖, 八股快问快答, 算法陪练, JD 定向面, 简历拷打, 复盘教练"},
                    "strength": {"type": "string", "description": "Interview strength: 人上人, 顶级, 夯, NPC, 拉完了"},
                    "tone": {"type": "string", "description": "Interviewer tone: 默认, 温和, 铁面"},
                    "level": {"type": "string", "description": "Difficulty level: 简单, 中等, 困难"},
                    "role": {"type": "string", "description": "Role bias: backend-java, backend-python, backend-go, ai-agent, ai-rag, ai-eval, sre-platform"},
                    "focus": {"type": "array", "items": {"type": "string"}, "description": "Focus topics"},
                },
                "required": [],
            },
        },
        {
            "name": "interview_session_jd",
            "description": "Set or update the job description for this interview session.",
            "input_schema": {
                "type": "object",
                "properties": {
                    "jd_text": {"type": "string", "description": "Full job description text"}
                },
                "required": ["jd_text"],
            },
        },
        {
            "name": "apply_llm_answer_judgement",
            "description": "Score a candidate's free-form answer and record it to the transcript. The judgement JSON is saved and the interview state advances automatically.",
            "input_schema": {
                "type": "object",
                "properties": {
                    "quality": {
                        "type": "string",
                        "enum": ["strong", "partial", "weak", "wrong"],
                        "description": "Answer quality level"
                    },
                    "score": {
                        "type": "integer",
                        "minimum": 1,
                        "maximum": 5,
                        "description": "Score 1-5"
                    },
                    "strengths": {
                        "type": "array",
                        "items": {"type": "string"},
                        "description": "What the candidate did well"
                    },
                    "issues": {
                        "type": "array",
                        "items": {"type": "string"},
                        "description": "What was missing or wrong"
                    },
                    "answer_summary": {
                        "type": "string",
                        "description": "Summary of the candidate's actual answer content"
                    },
                    "feedback": {
                        "type": "string",
                        "description": "One-line feedback to show the candidate"
                    },
                    "confidence": {
                        "type": "number",
                        "minimum": 0,
                        "maximum": 1,
                        "description": "Confidence in this judgement"
                    },
                    "next_followup": {
                        "type": "string",
                        "description": "Optional: next followup question if continuing"
                    },
                },
                "required": ["quality", "score", "answer_summary", "feedback"],
            },
        },
    ]


def execute_tool(tool_name: str, tool_input: dict[str, Any], session_dir: str) -> dict[str, Any]:
    """Execute a tool call by running the appropriate skill script."""
    sdir = session_manager._resolve_path(session_dir)
    start_time = time.time()

    try:
        if tool_name == "read_session_state":
            state = session_manager.read_session_state(sdir)
            return {
                "type": "tool_result",
                "tool_name": tool_name,
                "result": {
                    "runtime_status": state.get("runtime_status"),
                    "active_stage": state.get("active_stage"),
                    "stage_sequence": state.get("stage_sequence"),
                    "stage_status": state.get("stage_status"),
                    "config": state.get("config"),
                    "current_question": state.get("current_question"),
                    "progress": state.get("progress"),
                    "pending_reconfiguration": state.get("pending_reconfiguration"),
                },
            }

        elif tool_name == "interview_session_start":
            result = script_runner.run_script("interview_session.py", ["start", str(sdir)])
            after_state = session_manager.read_session_state(sdir)
            return {"type": "tool_result", "tool_name": tool_name, "result": result, "session_state": _state_snapshot(after_state)}

        elif tool_name == "interview_session_next":
            result = script_runner.run_script("interview_session.py", ["next", str(sdir)])
            after_state = session_manager.read_session_state(sdir)
            return {"type": "tool_result", "tool_name": tool_name, "result": result, "session_state": _state_snapshot(after_state)}

        elif tool_name == "interview_session_hint":
            result = script_runner.run_script("interview_session.py", ["hint", str(sdir)])
            after_state = session_manager.read_session_state(sdir)
            return {"type": "tool_result", "tool_name": tool_name, "result": result, "session_state": _state_snapshot(after_state)}

        elif tool_name == "interview_session_repeat":
            result = script_runner.run_script("interview_session.py", ["repeat", str(sdir)])
            after_state = session_manager.read_session_state(sdir)
            return {"type": "tool_result", "tool_name": tool_name, "result": result, "session_state": _state_snapshot(after_state)}

        elif tool_name == "interview_session_explain":
            result = script_runner.run_script("interview_session.py", ["explain", str(sdir)])
            after_state = session_manager.read_session_state(sdir)
            return {"type": "tool_result", "tool_name": tool_name, "result": result, "session_state": _state_snapshot(after_state)}

        elif tool_name == "interview_session_skip":
            args = ["skip", str(sdir)]
            if tool_input.get("feedback"):
                args.extend(["--feedback", tool_input["feedback"]])
            result = script_runner.run_script("interview_session.py", args)
            after_state = session_manager.read_session_state(sdir)
            return {"type": "tool_result", "tool_name": tool_name, "result": result, "session_state": _state_snapshot(after_state)}

        elif tool_name == "interview_session_pause":
            result = script_runner.run_script("interview_session.py", ["pause", str(sdir)])
            after_state = session_manager.read_session_state(sdir)
            return {"type": "tool_result", "tool_name": tool_name, "result": result, "session_state": _state_snapshot(after_state)}

        elif tool_name == "interview_session_continue":
            result = script_runner.run_script("interview_session.py", ["continue", str(sdir)])
            after_state = session_manager.read_session_state(sdir)
            return {"type": "tool_result", "tool_name": tool_name, "result": result, "session_state": _state_snapshot(after_state)}

        elif tool_name == "interview_session_score":
            result = script_runner.run_script("interview_session.py", ["score", str(sdir)])
            after_state = session_manager.read_session_state(sdir)
            return {"type": "tool_result", "tool_name": tool_name, "result": result, "session_state": _state_snapshot(after_state)}

        elif tool_name == "interview_session_report":
            result = script_runner.run_script("interview_session.py", ["report", str(sdir)])
            after_state = session_manager.read_session_state(sdir)
            eval_data = session_manager.read_evaluation(sdir)
            return {
                "type": "tool_result",
                "tool_name": tool_name,
                "result": result,
                "session_state": _state_snapshot(after_state),
                "evaluation": eval_data,
            }

        elif tool_name == "interview_session_configure":
            args = ["configure", str(sdir)]
            for key in ("mode", "strength", "tone", "level", "role"):
                if tool_input.get(key):
                    args.extend([f"--{key}", tool_input[key]])
            if tool_input.get("focus"):
                args.extend(["--focus", ",".join(tool_input["focus"])])
            result = script_runner.run_script("interview_session.py", args, timeout=60)
            after_state = session_manager.read_session_state(sdir)
            return {"type": "tool_result", "tool_name": tool_name, "result": result, "session_state": _state_snapshot(after_state)}

        elif tool_name == "interview_session_jd":
            args = ["jd", str(sdir), "--jd-text", tool_input.get("jd_text", "")]
            result = script_runner.run_script("interview_session.py", args)
            after_state = session_manager.read_session_state(sdir)
            return {"type": "tool_result", "tool_name": tool_name, "result": result, "session_state": _state_snapshot(after_state)}

        elif tool_name == "apply_llm_answer_judgement":
            # Save judgement to temp JSON, then call apply_llm_answer_judgement.py
            judgement = {
                "quality": tool_input.get("quality", "partial"),
                "score": tool_input.get("score", 3),
                "strengths": tool_input.get("strengths", []),
                "issues": tool_input.get("issues", []),
                "answer_summary": tool_input.get("answer_summary", ""),
                "feedback": tool_input.get("feedback", ""),
                "confidence": tool_input.get("confidence", 0.75),
                "next_followup": tool_input.get("next_followup", ""),
            }
            # Write to temp dir and call script
            with tempfile.NamedTemporaryFile(mode="w", suffix=".json", delete=False, encoding="utf-8") as f:
                json.dump(judgement, f, ensure_ascii=False)
                temp_path = f.name
            result = script_runner.run_script(
                "apply_llm_answer_judgement.py", [str(sdir), temp_path], timeout=30
            )
            # Clean up temp file (script may delete it if inside session dir)
            try:
                Path(temp_path).unlink(missing_ok=True)
            except OSError:
                pass
            after_state = session_manager.read_session_state(sdir)
            return {
                "type": "tool_result",
                "tool_name": tool_name,
                "result": result,
                "session_state": _state_snapshot(after_state),
            }

        else:
            return {"type": "tool_result", "tool_name": tool_name, "error": f"Unknown tool: {tool_name}"}

    except Exception as e:
        return {"type": "tool_result", "tool_name": tool_name, "error": str(e)}


def _state_snapshot(state: dict[str, Any]) -> dict[str, Any]:
    """Extract a minimal useful state snapshot."""
    return {
        "runtime_status": state.get("runtime_status"),
        "active_stage": state.get("active_stage"),
        "stage_status": state.get("stage_status"),
        "current_question": state.get("current_question"),
        "progress": state.get("progress"),
        "available_commands": state.get("available_commands"),
    }


def build_messages_from_transcript(
    transcript: dict[str, Any],
    state: dict[str, Any],
    user_message: str,
) -> list[dict[str, Any]]:
    """Build the Anthropic messages array from transcript history and current user input."""
    messages = []

    # Add context about current session state as a system-like user message
    state_context = _build_state_context(state)
    if state_context:
        messages.append({"role": "user", "content": state_context})

    # Add previous Q&A pairs from transcript
    answers = transcript.get("answers", [])
    for ans in answers[-20:]:  # Last 20 exchanges
        q_text = ans.get("question_text", "")
        if q_text:
            messages.append({"role": "assistant", "content": q_text})
        answer_text = ans.get("candidate_answer_text") or ans.get("candidate_answer_summary", "")
        if answer_text:
            messages.append({"role": "user", "content": answer_text})

        # Add judgement context if available
        quality = ans.get("quality")
        score_val = ans.get("score")
        if quality or score_val:
            fb = ans.get("interviewer_feedback", "")
            if fb:
                messages.append({
                    "role": "assistant",
                    "content": f"[Feedback: {fb} | Quality: {quality or 'N/A'} | Score: {score_val or 'N/A'}/5]",
                })

    # Add the current user input
    messages.append({"role": "user", "content": user_message})

    return messages


def _build_state_context(state: dict[str, Any]) -> str:
    """Build a concise state context string for the model."""
    runtime_status = state.get("runtime_status", "INIT")
    active_stage = state.get("active_stage")
    stage_status = state.get("stage_status", {})
    config = state.get("config", {})
    current_q = state.get("current_question") or {}
    progress = state.get("progress", {})
    pending = state.get("pending_reconfiguration") or {}

    lines = [f"[Session State: runtime_status={runtime_status}]"]

    if active_stage:
        lines.append(f"active_stage={active_stage}, stage_status={json.dumps(stage_status, ensure_ascii=False)}")

    if config:
        lines.append(f"config: mode={config.get('mode', 'N/A')}, strength={config.get('strength', 'N/A')}, tone={config.get('tone', 'N/A')}, level={config.get('level', 'N/A')}, role={config.get('role', 'N/A')}, focus={config.get('focus', [])}")

    if config.get("jd_text"):
        jd = config["jd_text"]
        lines.append(f"JD (full text): {jd}")

    if current_q and current_q.get("question_text"):
        lines.append(f"Current question: \"{current_q['question_text']}\" [stage={current_q.get('stage', '')}, hint_level={current_q.get('hint_level', 0)}]")

    if progress:
        lines.append(f"Progress: {json.dumps(progress, ensure_ascii=False)}")

    if pending:
        lines.append(f"Pending reconfiguration: {json.dumps(pending, ensure_ascii=False)}")

    lines.append("---")
    lines.append("You are the CS Technical Interviewer. Follow SKILL.md rules strictly. Use Chinese by default. Ask ONE question at a time. After the candidate answers, judge the answer using apply_llm_answer_judgement, then proceed to the next question.")

    return "\n".join(lines)


async def stream_chat_response(
    session_id: str,
    user_message: str,
) -> AsyncIterator[dict[str, Any]]:
    """Stream a full chat response with tool execution loop."""
    session_dir = session_manager._resolve_path(session_id)
    if not session_dir.exists():
        yield {"type": "error", "error": "Session not found"}
        return

    state = session_manager.read_session_state(session_dir)
    transcript = session_manager.read_transcript(session_dir)

    messages = build_messages_from_transcript(transcript, state, user_message)
    tools = build_tool_definitions()

    max_tool_rounds = 5
    tool_round = 0

    while tool_round < max_tool_rounds:
        tool_round += 1
        text_buffer = []
        tool_uses: list[dict[str, Any]] = []

        async for event in anthropic_client.stream_chat(
            system=SKILL_MD_CONTENT,
            messages=messages,
            tools=tools,
        ):
            event_type = event.get("type")

            if event_type == "text_delta":
                text_buffer.append(event.get("content", ""))
                yield {"type": "text_delta", "content": event.get("content", "")}

            elif event_type == "tool_use_start":
                tool_uses.append({
                    "tool_name": event.get("tool_name"),
                    "tool_id": event.get("tool_id"),
                    "tool_input": event.get("tool_input", {}),
                })
                yield {"type": "tool_use_start", "tool_name": event.get("tool_name")}

            elif event_type == "tool_use_delta":
                # Accumulate tool input JSON
                pass

            elif event_type == "message_done":
                break

            elif event_type == "error":
                yield event
                return

        # If no tool calls, we're done
        if not tool_uses:
            break

        # Execute tool calls
        for tool_use in tool_uses:
            tool_result = execute_tool(
                tool_use["tool_name"],
                tool_use["tool_input"],
                session_id,
            )
            yield {
                "type": "tool_result",
                "tool_name": tool_use["tool_name"],
                "result": tool_result,
            }

            # Add tool use + result to messages for next loop
            messages.append({
                "role": "assistant",
                "content": [
                    {
                        "type": "tool_use",
                        "id": tool_use.get("tool_id", "tool_1"),
                        "name": tool_use["tool_name"],
                        "input": tool_use["tool_input"],
                    }
                ],
            })
            messages.append({
                "role": "user",
                "content": [
                    {
                        "type": "tool_result",
                        "tool_use_id": tool_use.get("tool_id", "tool_1"),
                        "content": json.dumps(tool_result, ensure_ascii=False),
                    }
                ],
            })

    # Final state snapshot
    try:
        final_state = session_manager.read_session_state(session_dir)
        yield {
            "type": "message_done",
            "session_state": _state_snapshot(final_state),
            "full_state": final_state,
        }
    except FileNotFoundError:
        yield {"type": "message_done"}

File: frontend/src/services/chatService.ts
/**
 * Self-contained chat service — direct DeepSeek API calls with local tool execution.
 * Replaces: anthropic_client.py + chat_service.py + all Python scripts.
 */

import { getSkillMd } from './dataLoader';
import {
  saveSession, configureSession, recordAnswer, skipQuestion,
  hintQuestion, pauseInterview, continueInterview, generateReport,
} from './sessionManager';
import type { SessionState } from '../types/session';

// ========== DeepSeek API ==========

const API_KEY = 'sk-6397c0f9ee564435928af0b8052aaae2';
// Use proxy in browser (CORS), direct in native Capacitor WebView
const isNative = typeof (window as unknown as Record<string, unknown>).Capacitor !== 'undefined';
const API_BASE = isNative
  ? 'https://api.deepseek.com/anthropic'
  : '/api/deepseek/anthropic';
const MODEL = 'deepseek-v4-pro[1m]';

// ========== Tool Definitions (same logic as Python scripts, executed locally) ==========

const TOOLS = [
  {
    name: 'read_session_state',
    description: 'Read current interview session state. Call this first.',
    input_schema: { type: 'object', properties: {}, required: [] },
  },
  {
    name: 'interview_start',
    description: 'Start the interview — generate question plan and begin.',
    input_schema: { type: 'object', properties: {}, required: [] },
  },
  {
    name: 'apply_answer_judgement',
    description: 'Score candidate answer and advance to next question. MUST call after every candidate response.',
    input_schema: {
      type: 'object',
      properties: {
        quality: { type: 'string', enum: ['strong', 'partial', 'weak', 'wrong'] },
        score: { type: 'integer', minimum: 1, maximum: 5 },
        strengths: { type: 'array', items: { type: 'string' } },
        issues: { type: 'array', items: { type: 'string' } },
        answer_summary: { type: 'string' },
        feedback: { type: 'string' },
        confidence: { type: 'number', minimum: 0, maximum: 1 },
        next_followup: { type: 'string' },
      },
      required: ['quality', 'score', 'answer_summary', 'feedback'],
    },
  },
  {
    name: 'interview_hint',
    description: 'Increase hint level for current question.',
    input_schema: { type: 'object', properties: {}, required: [] },
  },
  {
    name: 'interview_skip',
    description: 'Skip current question.',
    input_schema: { type: 'object', properties: {}, required: [] },
  },
  {
    name: 'interview_pause',
    description: 'Pause the interview.',
    input_schema: { type: 'object', properties: {}, required: [] },
  },
  {
    name: 'interview_continue',
    description: 'Continue paused interview.',
    input_schema: { type: 'object', properties: {}, required: [] },
  },
  {
    name: 'interview_score',
    description: 'Generate mid-session score.',
    input_schema: { type: 'object', properties: {}, required: [] },
  },
  {
    name: 'interview_report',
    description: 'Generate final report.',
    input_schema: { type: 'object', properties: {}, required: [] },
  },
  {
    name: 'interview_configure',
    description: 'Update interview configuration.',
    input_schema: {
      type: 'object',
      properties: {
        mode: { type: 'string' },
        strength: { type: 'string' },
        tone: { type: 'string' },
        level: { type: 'string' },
        role: { type: 'string' },
        focus: { type: 'array', items: { type: 'string' } },
        jd_text: { type: 'string' },
      },
      required: [],
    },
  },
];

// ========== Context Builder ==========

function buildStateContext(state: SessionState): string {
  const { runtime_status, active_stage, stage_status, config, current_question, progress } = state;
  return [
    `[Session: ${runtime_status}]`,
    `active_stage=${active_stage || 'none'}`,
    `stage_status=${JSON.stringify(stage_status || {})}`,
    `config: mode=${config.mode}, strength=${config.strength}, tone=${config.tone}, level=${config.level}, role=${config.role || 'auto'}, focus=[${(config.focus || []).join(', ')}]`,
    config.jd_text ? `JD: ${config.jd_text.slice(0, 500)}` : '',
    state.question_plan && Object.keys(state.question_plan).length > 0
      ? `Question Plan: ${JSON.stringify(Object.fromEntries(Object.entries(state.question_plan).map(([k, v]) => [k, (v as unknown[]).length + ' questions'])))}`
      : '',
    current_question ? `Current: "${current_question.question_text}" [stage=${current_question.stage}, hint_level=${current_question.hint_level || 0}]` : 'No current question',
    active_stage && state.question_plan?.[active_stage]
      ? `Questions for ${active_stage}:\n${(state.question_plan[active_stage] as unknown[]).map((q: unknown, i: number) => `${i+1}. ${(q as Record<string,unknown>).question_id}: ${(q as Record<string,unknown>).question_text}`).join('\n')}`
      : '',
    `Progress: completed=${(progress.completed_question_ids || []).length}, hints=${progress.hints_used_total || 0}, skipped=${progress.skipped_total || 0}`,
    '---',
    'You are a CS Technical Interviewer. Follow SKILL.md rules strictly. Use Chinese. Ask ONE question at a time. After each answer call apply_answer_judgement then proceed.',
  ].join('\n');
}

// ========== Tool Execution ==========

interface ToolCallResult {
  success: boolean;
  session_state: SessionState;
  data?: unknown;
}

function executeTool(
  toolName: string,
  toolInput: Record<string, unknown>,
  state: SessionState,
): ToolCallResult {
  let newState = { ...state };

  switch (toolName) {
    case 'read_session_state':
      return { success: true, session_state: newState, data: stateSnapshot(newState) };

    case 'interview_start': {
      // Question plan will be generated asynchronously — for now set up the plan
      newState.runtime_status = 'RUNNING';
      if (newState.stage_sequence.length > 0) {
        newState.active_stage = newState.stage_sequence[0];
        newState.stage_status[newState.active_stage] = 'in_progress';
      }
      newState.command_history.push({ timestamp: new Date().toISOString(), command: 'start' });
      return { success: true, session_state: newState, data: stateSnapshot(newState) };
    }

    case 'apply_answer_judgement': {
      const quality = (toolInput.quality as string) || 'partial';
      const score = (toolInput.score as number) || 3;
      const summary = (toolInput.answer_summary as string) || '';
      const feedback = (toolInput.feedback as string) || '';
      newState = recordAnswer(newState, quality, score, summary, feedback);
      return { success: true, session_state: newState, data: { judged: true, quality, score } };
    }

    case 'interview_hint':
      newState = hintQuestion(newState);
      return { success: true, session_state: newState };

    case 'interview_skip':
      newState = skipQuestion(newState);
      return { success: true, session_state: newState };

    case 'interview_pause':
      newState = pauseInterview(newState);
      return { success: true, session_state: newState };

    case 'interview_continue':
      newState = continueInterview(newState);
      return { success: true, session_state: newState };

    case 'interview_score':
      newState.command_history.push({ timestamp: new Date().toISOString(), command: 'score' });
      return { success: true, session_state: newState, data: { score_pending: true } };

    case 'interview_report':
      newState = generateReport(newState);
      return { success: true, session_state: newState };

    case 'interview_configure': {
      newState = configureSession(newState, toolInput as Record<string, unknown>);
      return { success: true, session_state: newState };
    }

    default:
      return { success: false, session_state: newState, data: { error: `Unknown tool: ${toolName}` } };
  }
}

function stateSnapshot(state: SessionState): Record<string, unknown> {
  return {
    runtime_status: state.runtime_status,
    active_stage: state.active_stage,
    stage_status: state.stage_status,
    current_question: state.current_question,
    config: state.config,
    progress: state.progress,
  };
}

// ========== Streaming Chat ==========

export interface StreamEvents {
  onTextDelta: (text: string) => void;
  onToolCall: (name: string, input: Record<string, unknown>) => void;
  onToolResult: (name: string, result: ToolCallResult) => void;
  onError: (error: string) => void;
  onDone: (finalState: SessionState) => void;
}

export async function streamChat(
  state: SessionState,
  userMessage: string,
  events: StreamEvents,
): Promise<void> {
  const skillMd = await getSkillMd();
  const stateContext = buildStateContext(state);

  // Build messages
  const messages: Array<{ role: string; content: string }> = [
    { role: 'user', content: stateContext },
  ];

  // Add last few Q&A exchanges from state if available
  // (transcript would be stored separately)

  messages.push({ role: 'user', content: userMessage });

  // Call DeepSeek API
  try {
    const response = await fetch(`${API_BASE}/messages`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'x-api-key': API_KEY,
        'anthropic-version': '2023-06-01',
      },
      body: JSON.stringify({
        model: MODEL,
        max_tokens: 4096,
        system: [{ type: 'text', text: skillMd }],
        messages,
        tools: TOOLS,
        stream: true,
      }),
    });

    if (!response.ok) {
      const err = await response.text();
      events.onError(`API Error ${response.status}: ${err}`);
      return;
    }

    const reader = response.body?.getReader();
    if (!reader) { events.onError('No response stream'); return; }

    const decoder = new TextDecoder();
    let buffer = '';
    let currentState = { ...state };
    const pendingToolCalls: Array<{ id: string; name: string; input: string }> = [];
    let toolInputJson = '';

    while (true) {
      const { done, value } = await reader.read();
      if (done) break;

      buffer += decoder.decode(value, { stream: true });
      const lines = buffer.split('\n');
      buffer = lines.pop() || '';

      for (const line of lines) {
        if (!line.startsWith('data: ')) continue;
        const jsonStr = line.slice(6);
        if (jsonStr === '[DONE]') continue;

        try {
          const event = JSON.parse(jsonStr);

          if (event.type === 'content_block_delta') {
            const delta = event.delta;
            if (delta?.text) {
              events.onTextDelta(delta.text);
            }
            if (delta?.partial_json) {
              toolInputJson += delta.partial_json;
            }
          }

          if (event.type === 'content_block_start') {
            const block = event.content_block;
            if (block?.type === 'tool_use') {
              pendingToolCalls.push({ id: block.id, name: block.name, input: '' });
              toolInputJson = '';
            }
          }

          if (event.type === 'content_block_stop') {
            // Finalize tool call input
            if (pendingToolCalls.length > 0) {
              const last = pendingToolCalls[pendingToolCalls.length - 1];
              if (toolInputJson && !last.input) {
                last.input = toolInputJson;
              }
            }
          }

          if (event.type === 'message_delta' && event.delta?.stop_reason === 'tool_use') {
            // All tool calls finalized — execute them
            for (const tc of pendingToolCalls) {
              let parsedInput: Record<string, unknown> = {};
              try { parsedInput = JSON.parse(tc.input || '{}'); } catch { /* empty */ }

              events.onToolCall(tc.name, parsedInput);
              const result = executeTool(tc.name, parsedInput, currentState);
              currentState = result.session_state;
              events.onToolResult(tc.name, result);

              // Persist state after each tool execution
              await saveSession(currentState);
            }
            pendingToolCalls.length = 0;
            toolInputJson = '';
          }
        } catch {
          // Skip malformed events
        }
      }
    }

    // Auto-judge: if the LLM didn't call apply_answer_judgement, force it
    const isCommand = userMessage.startsWith('/') || /^(开始|暂停|继续|跳过|提示|重复|解释|评分|报告)/.test(userMessage);
    const hasQuestion = !!currentState.current_question;
    if (!isCommand && hasQuestion && !pendingToolCalls.length) {
      // Advance state with default judgement
      currentState = recordAnswer(currentState, 'partial', 3, userMessage.slice(0, 100), '');
      await saveSession(currentState);

      // Tell LLM to ask the next question
      if (currentState.current_question) {
        const ctx = buildStateContext(currentState);
        messages.push({ role: 'user', content: `[State updated. Ask the next question.]\n${ctx}` });
        // Re-call API for the next question
        const resp2 = await fetch(`${API_BASE}/messages`, {
          method: 'POST',
          headers: { 'Content-Type': 'application/json', 'x-api-key': API_KEY, 'anthropic-version': '2023-06-01' },
          body: JSON.stringify({ model: MODEL, max_tokens: 2048, system: [{ type: 'text', text: getSkillMd() }], messages, tools: TOOLS, stream: true }),
        });
        if (resp2.ok) {
          const reader2 = resp2.body?.getReader();
          if (reader2) {
            let buf2 = '';
            while (true) {
              const { done, value } = await reader2.read();
              if (done) break;
              buf2 += new TextDecoder().decode(value, { stream: true });
              for (const l of buf2.split('\n')) {
                if (l.startsWith('data: ') && l.slice(6) !== '[DONE]') {
                  try {
                    const evt = JSON.parse(l.slice(6));
                    if (evt.type === 'content_block_delta' && evt.delta?.text) {
                      events.onTextDelta(evt.delta.text);
                    }
                  } catch { /* skip */ }
                }
              }
              buf2 = buf2.split('\n').pop() || '';
            }
          }
        }
      }
    }

    // Final save
    await saveSession(currentState);
    events.onDone(currentState);
  } catch (err: unknown) {
    events.onError(err instanceof Error ? err.message : 'Unknown error');
  }
}

// ========== Non-streaming (for commands) ==========

export async function sendChat(
  state: SessionState,
  userMessage: string,
): Promise<{ text: string; state: SessionState }> {
  return new Promise((resolve, reject) => {
    let text = '';

    streamChat(state, userMessage, {
      onTextDelta: (t) => { text += t; },
      onToolCall: () => {},
      onToolResult: () => { /* state updated via onDone */ },
      onError: (err) => reject(new Error(err)),
      onDone: (s) => resolve({ text, state: s }),
    });
  });
}
